shared-module/display/ili9341.c

```diff
@@ -0,0 +1,229 @@
+    write_data(self, 0x03);
+    write_data(self, 0x11);
+    write_data(self, 0x07);
+    write_data(self, 0x31);
+    write_data(self, 0xC1);
+    write_data(self, 0x48);
+    write_data(self, 0x08);
+    write_data(self, 0x0F);
+    write_data(self, 0x0C);
+    write_data(self, 0x31);
+    write_data(self, 0x36);
+    write_data(self, 0x0F);
+
+    write_command(self, ILI9341_SLPOUT);
+    mp_hal_delay_ms(120);
+
+    write_command(self, ILI9341_DISPON);
+}
+
+void display_ili9341_set_rotation(ili9341_obj_t *self,
+        display_rotation_t rotation)
+{
+    uint8_t data;
+
+    self->rotation = rotation;
+
+    write_command(self, ILI9341_MADCTL);
+
+    switch (self->rotation) {
+        case ROTATION_90:  data = (ILI9341_MADCTL_MV | ILI9341_MADCTL_BGR); break;
+        case ROTATION_180: data = (ILI9341_MADCTL_MY | ILI9341_MADCTL_BGR); break;
+        case ROTATION_270: data = (ILI9341_MADCTL_MX | ILI9341_MADCTL_MY | ILI9341_MADCTL_MV | ILI9341_MADCTL_BGR); break;
+        default:           data = (ILI9341_MADCTL_MX | ILI9341_MADCTL_BGR); break;
+    }
+
+    write_data(self, data);
+}
+
+void display_ili9341_set_addr_window(ili9341_obj_t *self, uint16_t x_0,
+        uint16_t y_0, uint16_t x_1, uint16_t y_1) {
+    write_command(self, ILI9341_CASET);
+    write_data(self, x_0 >> 8);
+    write_data(self, x_0);
+    write_data(self, x_1 >> 8);
+    write_data(self, x_1);
+
+    write_command(self, ILI9341_PASET);
+    write_data(self, y_0 >> 8);
+    write_data(self, y_0);
+    write_data(self, y_1 >> 8);
+    write_data(self, y_1);
+
+    write_command(self, ILI9341_RAMWR);
+}
+
+void display_ili9341_pixel(ili9341_obj_t *self, uint16_t x, uint16_t y,
+        uint32_t color) {
+    const uint8_t data[2] = {color >> 8, color};
+
+    display_ili9341_set_addr_window(self, x, y, x, y);
+
+    common_hal_digitalio_digitalinout_set_value(self->dc, true);
+
+    spi_write(self, data, sizeof(data));
+
+    common_hal_digitalio_digitalinout_set_value(self->dc, false);
+}
+
+void display_ili9341_rect(ili9341_obj_t *self, uint16_t x, uint16_t y,
+        uint16_t width, uint16_t height, uint32_t color)
+{
+    const uint8_t data[2] = { color >> 8, color };
+
+    display_ili9341_set_addr_window(self, x, y, x + width - 1, y + height - 1);
+
+    common_hal_digitalio_digitalinout_set_value(self->dc, true);
+
+    // Duff's device algorithm
+    uint32_t i = (height * width + 7) / 8;
+
+    switch ((height * width) % 8) {
+        case 0: do { spi_write(self, data, sizeof(data));
+        case 7:      spi_write(self, data, sizeof(data));
+        case 6:      spi_write(self, data, sizeof(data));
+        case 5:      spi_write(self, data, sizeof(data));
+        case 4:      spi_write(self, data, sizeof(data));
+        case 3:      spi_write(self, data, sizeof(data));
+        case 2:      spi_write(self, data, sizeof(data));
+        case 1:      spi_write(self, data, sizeof(data));
+                } while (--i > 0);
+    }
```

Comment:
Do we really do that? I would guess that the speed of `spi_write` trumps the speed of the C code anyways.